src/components/structures/TimelinePanel.tsx

```diff
@@ -369,9 +396,74 @@ class TimelinePanel extends React.Component<IProps, IState> {
             client.removeListener(MatrixEventEvent.Replaced, this.onEventReplaced);
             client.removeListener(MatrixEventEvent.VisibilityChange, this.onEventVisibilityChange);
             client.removeListener(ClientEvent.Sync, this.onSync);
+            client.removeListener(ClientEvent.DumpDebugLogs, this.onDumpDebugLogs);
         }
     }
 
+    /**
+     * Logs out debug info to describe the state of the TimelinePanel and the
+     * events in the room according to the matrix-js-sdk. This is useful when
+     * debugging problems like messages out of order, or messages that should
+     * not be showing up in a thread, etc.
+     *
+     * It's too expensive and cumbersome to do all of these calculations for
+     * every message change so instead we only log it out when the
+     * BugReportDialog is opened.
+     */
+    private onDumpDebugLogs = (): void => {
+        const roomId = this.props.timelineSet.room.roomId;
+        // Get a list of the event IDs given to the TimelinePanel.
+        // This includes state and hidden events which we don't render
+        const eventIdList = this.state.events.map((ev) => ev.getId());
+
+        // Get the list of actually rendered events seen in the DOM.
+        // This is useful to know for sure what's being shown on screen.
+        // And we can suss out any corrupted React `key` problems.
+        let renderedEventIds;
+        const messagePanel = this.messagePanel.current;
+        if (messagePanel) {
+            const messagePanelNode = ReactDOM.findDOMNode(messagePanel) as HTMLElement;
+            if (messagePanelNode) {
+                const actuallyRenderedEvents = messagePanelNode.querySelectorAll('[data-event-id]');
+                renderedEventIds = [...actuallyRenderedEvents].map((renderedEvent) => {
+                    return renderedEvent.getAttribute('data-txn-id') || renderedEvent.getAttribute('data-event-id');
```

Comment:
The transaction ID is not an event ID though?

Reply:
🤷‍♀️ I was trying to mimic what the `key` is and account for local echo events which are also important to know about, [`src/components/structures/MessagePanel.tsx#L793`](https://github.com/matrix-org/matrix-react-sdk/blob/e050d8432acca299e7050102f906a84ae0c1160e/src/components/structures/MessagePanel.tsx#L793)

I think we can tell the difference by looking at it?

Better name?

Reply:
Given that local echo events return something for `event.getId()`, ex. `~!ifHvTosMVvhbsCtKis:my.synapse.server:m1651798905781.0`, I think we can drop this distinction since we can always derive the direct `txnId` if we want from that value (`m1651798905781.0`)

⏩